packages/shopping-cart/src/use-shopping-cart-manager.ts

```diff
@@ -122,10 +128,10 @@ export default function useShoppingCartManager( {
 			replaceProductInCart,
 			replaceProductsInCart,
 			reloadFromServer,
-			responseCart: responseCartWithoutTempProducts,
+			responseCart: lastValidResponseCart.current,
 		} ),
 		[
-			responseCartWithoutTempProducts,
+			lastValidResponseCart,
```

Comment:
@sirbrillig This `lastValidResponseCart` dependency could be a bug. It's a ref and therefore it's always the same object. Only the `.current` field ever changes. Changing `lastValidResponseCart` will never trigger an update of `shoppingCartManager` and we might return a `shoppingCartManager` with a stale `.responseCart` field.

It probably works anyway because every relevant cart change is accompanied with a change of `isPendingUpdate`?

Another case of the same issue in the same file is when the `isMounted` ref is used as a dependency of an `useCallback` hook. It doesn't do any harm. It's just a redundant dependency that might be interpreted as a (false) promise that something will get updated on `isMounted` change. But it won't.

Reply:
Oh, good point! 🤦 That said, this is a little tricky to do because:

As you said, if we don't depend on a non-ref version of the cart, there's a risk (however small) that the `responseCart` returned by the hook will have out-of-date data.

But we don't want to return the updated data if it's invalid (the point of this PR).

But we also don't want to create a new `useState` for the valid cache (this is how it was several versions ago) because then updating the cache (with a `useEffect`) will require another render before `responseCart` is correct, and during that render the other variables (particularly `isPendingUpdate`) will have already updated, providing misleading data to the consumer.

I think the best solution might be to restore the dependency of `responseCartWithoutTempProducts` here, even though it's not being used in the memoized value. That way we will force `responseCart` to always be up-to-date. I've tried to do that in https://github.com/Automattic/wp-calypso/pull/54560

Reply:
> It probably works anyway because every relevant cart change is accompanied with a change of isPendingUpdate?

More specifically, any change to the cart will trigger several changes to `cacheStatus` (the dependency for changing `lastValidResponseCart`), which then causes a change to one of the other variables that is already a dependency: `isLoading`, `isPendingUpdate`, or `loadingErrorForManager`. I'm pretty sure there's no way currently for a cart update without that happening but it's worth considering.

Reply:
> I tried depending on `lastValidResponseCart.current`, but that gives an eslint warning and I'm not 100% sure if it would work.

Depending on `lastValidResponseCart.current` is the right thing to do here. Depending on `responseCartWithoutTempProducts` is not bad either, just inefficient: it will trigger unneeded updates of `shoppingCartManager` when `cacheStatus` is not `valid`.

The `react-hooks/exhaustive-deps` rule complains about any dependency that has the `*.current` format because refs can be unreliable dependencies: they can be updated by anyone at any time and don't guarantee a rerender.

In our case though, `lastValidResponseCart` depends on two bits of state: `cacheStatus` and `responseCartWithoutTempProducts`, and nothing else. Rerenders are guaranteed by the underlying state changes.

You just need to trick the lint rule by assigning the value to a variable first (`const responseCart = lVRC.current`) and then using the variable as dependency. The lint rule doesn't understand your code -- it just applies a set of fallible heuristics.